File: python/907.py
from typing import List


class Solution:
    def sumSubarrayMins(self, arr: List[int]) -> int:
        n = len(arr)
        left = [-1] * n
        st = [-1]
        for i in range(n):
            while len(st) > 1 and st[-1] >= arr[i]:
                st.pop()
            left[i] = st[-1]
            st.append(i)
        right = [-1] * n
        st = [n]
        for i in range(n, -1):
            while len(st) > 0 and st[-1] > arr[i]:
                st.pop()
            right[i] = st[-1]
            st.append(i)
        ans = 0
        for i in range(n):
            ans += (right[i] - i) * (i - left[i]) * arr[i]
        return ans % (1e9 + 7)

File: go/907.go
package main

func sumSubarrayMins(arr []int) int {
	n := len(arr)
	left := make([]int, n)
	st := []int{-1}
	for i, num := range arr {
		for len(st) > 1 && arr[st[len(st)-1]] >= num {
			st = st[:len(st)-1]
		}
		left[i] = st[len(st)-1]
		st = append(st, i)
	}
	right := make([]int, n)
	st = []int{n}
	for i := n - 1; i >= 0; i-- {
		for len(st) > 1 && arr[st[len(st)-1]] > arr[i] {
			st = st[:len(st)-1]
		}
		right[i] = st[len(st)-1]
		st = append(st, i)
	}
	ans := 0
	for i, num := range arr {
		ans += (i - left[i]) * (right[i] - i) * num
	}
	return ans % (1_000_000_000 + 7)
}
